addons/test_mail_full/tests/test_portal.py

```diff
@@ -145,6 +147,188 @@ def test_portal_share_comment(self):
+    through the embedded link:
+    - either in the backend if the user is connected and has the right to
+    - or in the portal otherwise
+    """
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.customer = cls.env['res.partner'].create({
+            'country_id': cls.env.ref('base.be').id,
+            'email': 'mdelvaux34@example.com',
+            'lang': 'en_US',
+            'mobile': '+32492131422',
+            'name': 'Mathias Delvaux',
+            'phone': '+3227411823',
+        })
+        cls.record_portal = cls.env['mail.test.portal'].create({
+            'name': 'Test Portal Record',
+            'partner_id': cls.customer.id,
+            'user_id': cls.user_admin.id,
+        })
+        cls.mail_template = cls.env['mail.template'].create({
+            'auto_delete': True,
+            'body_html': '<p>Hello <t t-out="object.partner_id.name"/>, your quotation is ready for review.</p>',
+            'email_from': '{{ (object.user_id.email_formatted or user.email_formatted) }}',
+            'model_id': cls.env.ref('test_mail_full.model_mail_test_portal').id,
+            'name': 'Quotation template',
+            'partner_to': '{{ object.partner_id.id }}',
+            'subject': 'Your quotation "{{ object.name }}"',
+        })
+        cls._create_portal_user()
+        for group_name, group_func, group_data in cls.record_portal.sudo()._notify_get_recipients_groups():
+            if group_name == 'portal_customer' and group_func(cls.customer):
+                cls.record_access_url = group_data['button_access']['url']
+                break
+        else:
+            raise AssertionError('Record access URL not found')
+        # Build record_access_url_wrong_token with altered access_token for testing security
+        parsed_url = url_parse(cls.record_access_url)
+        query_params = url_decode(parsed_url.query)
+        cls.record_access_url_wrong_token = url_unparse(
+            (parsed_url[0], parsed_url[1], parsed_url[2],
+             url_encode({**query_params,
+                         'access_token': query_params['access_token'].translate(
+                             str.maketrans('0123456789abcdef',
+                                           '9876543210fedcba'))},
+                        sort=True),
+             parsed_url[4]))
+
+    def _assert_URL(self, url, path, expected_fragment_params=None, expected_in_query=None):
+        parsed_url = url_parse(url)
+        fragment_params = url_decode(parsed_url.fragment)
+        self.assertEqual(parsed_url.path, path, 'URL path')
+        if expected_fragment_params:
+            for key, value in expected_fragment_params.items():
+                self.assertEqual(fragment_params.get(key), value, f'Param: {key}')
+        if expected_in_query:
+            self.assertIn(expected_in_query, parsed_url.query, f'Query of {path}')
```

Comment:
If you add a message, I would add the expected versus actual in it, to ease understand the fail (same for above) otherwise you still have to add a print when it fails :D

Reply:
he idea was to give the context in the message but yes, it is better to set the expected value as well as the context. Here is how it displays now: 

AssertionError: 'mail.test.portal' != 'mail.test.portal-'
- mail.test.portal
+ mail.test.portal-
?                 +
 : Expected: "model=mail.test.portal" (for path: /web/login)